odoo/orm/models.py

```diff
@@ -3485,12 +3488,15 @@ def fetch(self, field_names: Collection[str]) -> None:
 
         # fetch the fields
         fetched = self._fetch_query(query, fields_to_fetch)
+        env = self.env
+        if not env.su and self._check_access is BaseModel._check_access:
+            env.transaction._add_to_access_cache(fetched)
 
         # possibly raise exception for the records that could not be read
         if fetched != self:
             forbidden = (self - fetched).exists()
             if forbidden:
-                raise self.env['ir.rule']._make_access_error('read', forbidden)
+                raise env['ir.rule']._make_access_error('read', forbidden)
```

Comment:
This code actually doesn't make sense in `sudo()` mode, since `forbidden` will always be empy.  We could save one query when checking ir.rules, since the domain is evaluated in `sudo()` mode.